File: 백준/Silver/14501. 퇴사/퇴사.py
# 퇴사

n = int(input())
arr = [list(map(int, input().split())) for i in range(n)]
dp = [0] * (n+1)

val = 0
for i in range(n-1, -1, -1):
    if i + arr[i][0] > n:
        if i==n-1:
            dp[i]=0
        else:
            dp[i]=max(0, dp[i+1])
        continue
    dp[i] = max(dp[i+1], dp[i+arr[i][0]]+arr[i][1])

print(dp[0])

File: 백준/Silver/14501. 퇴사/퇴사.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        int[][] arr = new int[n + 1][2];
        StringTokenizer st;
        int[] dp = new int[n + 2];
        for (int i = 1; i < n+1; i++) {
            st = new StringTokenizer(br.readLine());
            arr[i][0] = Integer.parseInt(st.nextToken());
            arr[i][1] = Integer.parseInt(st.nextToken());
        }

        int tmp;
        int answer = 0;
        int max = 0;
        for (int i = 1; i < n + 1; i++) {
            tmp = i + arr[i][0];
            if (tmp > n+1) continue;
            dp[i] = Math.max(dp[i-1], dp[i]);
            dp[tmp] = Math.max(dp[i]+arr[i][1], dp[tmp]);
            answer = Math.max(answer, dp[tmp]);
//            max = Math.max(dp[i], max);
        }

//        for(int a: dp)
//            System.out.print(a+" ");
//        System.out.println();
        System.out.println(answer);
//        System.out.println(dp[n]);
    }
}